pkg/console/service/kubevirt-proxy.go

```diff
@@ -1,29 +1,253 @@
+	return &kubevirtProxy{
+		kubevirtClient: kubevirtClient,
+	}
 }
 
 type kubevirtProxy struct {
+	kubevirtClient kubecli.KubevirtClient
 }
 
 var _ Service = &kubevirtProxy{}
 
 func (k *kubevirtProxy) TokenHandler(request *restful.Request, response *restful.Response) {
-	//TODO: implement
-	panic("not implemented")
+	err := ValidateVncRbac(request, k.kubevirtClient)
+	if err != nil {
+		_ = response.WriteError(http.StatusUnauthorized, err)
+		return
+	}
+
+	namespace := request.PathParameter("namespace")
+	name := request.PathParameter("name")
+
+	duration := defaultTokenDuration
+	durationParam := request.QueryParameter("duration")
+	if durationParam != "" {
+		var err error
+		duration, err = time.ParseDuration(durationParam)
+		if err != nil {
+			_ = response.WriteError(http.StatusBadRequest, fmt.Errorf("failed to parse duration: %w", err))
+			return
+		}
+	}
+
+	vm, err := k.kubevirtClient.VirtualMachine(namespace).Get(name, &metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_ = response.WriteError(http.StatusNotFound, fmt.Errorf("VirtualMachine does not exist: %w", err))
+			return
+		}
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("error getting VirtualMachine: %w", err))
+		return
+	}
+
+	serviceAccountName := vm.Name + "-vnc-access"
+	err = k.createResources(request.Request.Context(), serviceAccountName, vm)
+	if err != nil {
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("error creating resources: %w", err))
+		return
+	}
+
+	durationSeconds := int64(duration.Seconds())
+	tokenRequest := &authnv1.TokenRequest{
+		Spec: authnv1.TokenRequestSpec{
+			Audiences:         nil,
+			ExpirationSeconds: &durationSeconds,
+			BoundObjectRef:    nil,
+		},
+	}
+
+	tokenRequest, err = k.kubevirtClient.CoreV1().ServiceAccounts(namespace).CreateToken(
+		request.Request.Context(),
+		serviceAccountName,
+		tokenRequest,
+		metav1.CreateOptions{},
+	)
+	if err != nil {
+		_ = response.WriteError(http.StatusInternalServerError, fmt.Errorf("failed to create token: %w", err))
+		return
+	}
+
+	_ = response.WriteAsJson(&v1alpha1.TokenResponse{
+		Token: tokenRequest.Status.Token,
+	})
 }
 
-func (k *kubevirtProxy) VncHandler(request *restful.Request, response *restful.Response) {
-	//TODO: implement
-	panic("not implemented")
+func (k *kubevirtProxy) VncHandler(_ *restful.Request, response *restful.Response) {
+	_ = response.WriteError(http.StatusNotFound, fmt.Errorf(
+		"/vnc endpoint is not available, when the VM console proxy is configured to use kubernetes API to generate tokens",
+	))
+}
+
+func (k *kubevirtProxy) createResources(ctx context.Context, name string, vmMeta metav1.Object) error {
+	namespace := vmMeta.GetNamespace()
+	commonLabels := map[string]string{
+		AppKubernetesNameLabel:      "vm-console-proxy",
+		AppKubernetesPartOfLabel:    "vm-console-proxy",
+		AppKubernetesManagedByLabel: "vm-console-proxy",
+		AppKubernetesComponentLabel: "vm-console-proxy",
```

Comment:
Can we use constant string that will be stored in `constants.go`?

Reply:
I've extracted it to a constant, but I didn't to move it to `constants.go`. That file contains constants that are used in multiple other files.